src/txmempool.h

```diff
@@ -549,8 +573,11 @@ class CTxMemPool
 
     std::vector<indexed_transaction_set::const_iterator> GetSortedDepthAndScore() const EXCLUSIVE_LOCKS_REQUIRED(cs);
 
-    /** track locally submitted transactions to periodically retry initial broadcast */
-    std::set<uint256> m_unbroadcast_txids GUARDED_BY(cs);
+    /**
+     * track locally submitted transactions to periodically retry initial broadcast
+     * map of txid -> wtxid
+     */
+    std::map<uint256, uint256> m_unbroadcast_txids GUARDED_BY(cs);
```

Comment:
I don't think this change is necessary (or desirable). I guess it was prompted by this comment: https://github.com/bitcoin/bitcoin/pull/18038#discussion_r393293074 , but storing this as a map from txid->wtxid is unnecessary and confusing, since it obfuscates the purpose of this data, which is a *set* of transactions that may need to be rebroadcast.

The only place that the wtxid is read from this map is in `ReattemptInitialBroadcast()`, at which point we can fetch the transaction from the mempool (we very nearly do that with `m_mempool.exists()`), and get the wtxid from there.

cc @amitiuttarwar @ariard

Reply:
My mistake here, effectively you need to announce transactions from the unbroadcast set by wtxid to upgraded peers but as you point out it doesn't mean we need to cache them.

Maybe we should precise unbroadcast set semantic wrt same txid, different wtxids, we only guarantee reattempt to the best mempool candidate known for a given txid at the time we call `ReattemptInitialBroadcast`, not insertion. https://github.com/bitcoin/bitcoin/pull/18044#discussion_r458229569 isn't an issue.

Reply:
I really think this "same txid, different wtxid" thing is a complete red herring. The mempool can only ever be aware of one witness for a transaction, so any attempt to announce the transaction via a different wtxid would fail anyway. As Suhas pointed out earlier (https://github.com/bitcoin/bitcoin/pull/18044#discussion_r458383811), to support tracking multiple witnesses we'd need to make significant changes many of our components. Besides, the unbroadcast set is transactions that we created ourselves, so we're not expecting the witnesses to change.

Reply:
Agree with @jnewbery. There is no real way that we can reasonable start tracking multiple witnesses for the same transaction (either in the mempool or the wallet, ...), so unless this map's storing of wtxid is needed for efficiency, it seems just a set of txid should be sufficient here.

Reply:
😮   wow that's a great point @jnewbery. thank you!!

I thought during implementation I ran into something that prevented me from exclusively having txids on the set, but that evaluation is simply wrong. I've taken a first pass at reverting unbroadcast to a set here: https://github.com/amitiuttarwar/bitcoin/commit/f51cccdd6c97d1b1e45163d56ebd6b7d29a2f587. I'll need to revisit with fresh eyes and then will PR. 

I agree set makes more sense than a map- its simpler, communicates the intent better, and there's no noticeable efficiency gain with the map. Since we check the transaction is the mempool, the difference between map vs set is the difference between calling `CTxMemPool::get` vs `CTxMemPool::exists`. This boils down to `mapTx.find(hash)` vs `mapTx.count(hash)`, which both have log(n) complexity (according to the boost docs I found).

Reply:
@amitiuttarwar mapTX is a multiindex and we use a hashed index not a ordered index so it's O(1) lookup. Not sure exactly how mapTX relates to m_unboardcast_txids though...

Reply:
@amitiuttarwar that patch looks correct from a quick skim. I'll review fully once you open a PR.

As @JeremyRubin points out, we're using a `boost::multi_index::hashed_unique` index for the wtxid index, which has constant time search/retrieval.

Reply:
reverted to set in https://github.com/bitcoin/bitcoin/pull/19879/commits/cb79b9dbf4cd06e17c8c65b36bf15c3ea2641de4, #19879